File: src/main/java/com/kingdom/util/computercardaction/KingdomComputerCardActionHandler.java
package com.kingdom.util.computercardaction;

import com.kingdom.model.Card;
import com.kingdom.model.CardAction;
import com.kingdom.model.Game;
import com.kingdom.model.Player;
import com.kingdom.model.computer.ComputerPlayer;
import com.kingdom.util.cardaction.CardActionHandler;

import java.util.ArrayList;
import java.util.List;

public class KingdomComputerCardActionHandler {
    public static void handleCardAction(CardAction cardAction, ComputerPlayer computer) {

        Game game = computer.getGame();
        Player player = computer.getPlayer();

        String cardName = cardAction.getCardName();
        int type = cardAction.getType();

        switch (cardName) {
            case "Artisan":
                if (cardAction.getType() == CardAction.TYPE_GAIN_CARDS_FROM_SUPPLY) {
                    chooseHighestCostCard(cardAction, computer, game, player);
                } else {
                    if (player.getActions() == 0 && !player.getActionCards().isEmpty()) {
                        Card card = computer.getHighestCostCard(player.getActionCards());
                        chooseCard(cardAction, game, player, card);
                    } else {
                        chooseLowestCostCard(cardAction, computer, game, player);
                    }
                }
                break;
            case "Bureaucrat": {
                List<Integer> cardIds = computer.getCardsNotNeeded(cardAction.getCards(), 1);
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
                break;
            }
            case "Cellar":
                List<Integer> cardsToDiscard = new ArrayList<>();
                for (Card card : cardAction.getCards()) {
                    if (computer.isCardToDiscard(card)) {
                        cardsToDiscard.add(card.getCardId());
                    }
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardsToDiscard, null, null, -1);
                break;
            case "Chancellor":
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, "yes", null, -1);
                break;
            case "Chapel":
                List<Integer> cardsToTrash = new ArrayList<>();
                for (Card card : cardAction.getCards()) {
                    if (computer.isCardToTrash(card)) {
                        cardsToTrash.add(card.getCardId());
                    }
                    if (cardsToTrash.size() == 4) {
                        break;
                    }
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardsToTrash, null, null, -1);
                break;
            case "Feast":
                chooseHighestCostCard(cardAction, computer, game, player);
                break;
            case "Library":
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, "yes", null, -1);
                break;
            case "Militia":
                int numCardsToDiscard = player.getHand().size() - 3;
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, computer.getCardsToDiscard(cardAction.getCards(), numCardsToDiscard), null, null, -1);
                break;
            case "Mine":
                if (type == CardAction.TYPE_TRASH_CARDS_FROM_HAND) {
                    chooseLowestCostCard(cardAction, computer, game, player);
                } else {
                    chooseHighestCostCard(cardAction, computer, game, player);
                }
                break;
            case "Remodel":
                if (type == CardAction.TYPE_TRASH_CARDS_FROM_HAND) {
                    chooseLowestCostCard(cardAction, computer, game, player);
                } else {
                    chooseHighestCostCard(cardAction, computer, game, player);
                }
                break;
            case "Spy":
                String yesNoAnswer = "yes";
                Card topCard = cardAction.getCards().get(0);
                if (topCard.getCardId() == Card.CURSE_ID || topCard.getCardId() == Card.COPPER_ID) {
                    yesNoAnswer = "no";
                }
                for (Card card : player.getVictoryCards()) {
                    if (!card.isTreasure() && !card.isAction()) {
                        yesNoAnswer = "no";
                    }
                }
                if (cardAction.getPlayerId() == player.getUserId()) {
                    if (yesNoAnswer.equals("yes")) {
                        yesNoAnswer = "no";
                    } else {
                        yesNoAnswer = "yes";
                    }
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, yesNoAnswer, null, -1);
                break;
            case "Thief": {
                List<Integer> cardIds = new ArrayList<>();
                if (type == CardAction.TYPE_CHOOSE_CARDS) {
                    if (cardAction.getNumCards() >= 1) {
                        Card card1 = cardAction.getCards().get(0);
                        if (cardAction.getCards().size() == 2) {
                            Card card2 = cardAction.getCards().get(1);
                            if (card1.isTreasure() && card2.isTreasure()) {
                                if (card1.getCost() > card2.getCost()) {
                                    cardIds.add(card1.getCardId());
                                } else {
                                    cardIds.add(card2.getCardId());
                                }
                            } else if (card1.isTreasure()) {
                                cardIds.add(card1.getCardId());
                            } else {
                                cardIds.add(card2.getCardId());
                            }
                        } else {
                            cardIds.add(card1.getCardId());
                        }
                    }
                } else {
                    if (cardAction.getNumCards() > 0) {
                        for (Card card : cardAction.getCards()) {
                            if (card.getCost() > 0) {
                                cardIds.add(card.getCardId());
                            }
                        }
                    }
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
                break;
            }
            case "Throne Room": {
                Card cardToPlay = computer.getActionToDuplicate(cardAction.getCards(), 2);
                chooseCard(cardAction, game, player, cardToPlay);
                break;
            }
            case "Workshop": {
                Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
                List<Integer> cardIds = new ArrayList<>();
                cardIds.add(cardToGain.getCardId());
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
                break;
            }
            default:
                throw new RuntimeException("Kingdom Card Action not handled for card: " + cardAction.getCardName() + " and type: " + cardAction.getType());
        }
    }

    private static void chooseLowestCostCard(CardAction cardAction, ComputerPlayer computer, Game game, Player player) {
        Card card = computer.getLowestCostCard(cardAction.getCards());
        chooseCard(cardAction, game, player, card);
    }

    private static void chooseHighestCostCard(CardAction cardAction, ComputerPlayer computer, Game game, Player player) {
        Card card = computer.getHighestCostCard(cardAction.getCards());
        chooseCard(cardAction, game, player, card);
    }

    private static void chooseCard(CardAction cardAction, Game game, Player player, Card card) {
        if (card == null) {
            card = cardAction.getCards().get(0);
        }
        List<Integer> cardIds = new ArrayList<>();
        cardIds.add(card.getCardId());
        CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
    }
}

File: src/main/kotlin/com/kingdom/util/computercardaction/KingdomComputerCardActionHandler.kt
package com.kingdom.util.computercardaction

import com.kingdom.model.cards.Card
import com.kingdom.model.OldCardAction
import com.kingdom.model.Game
import com.kingdom.model.Player
import com.kingdom.model.cards.supply.Curse
import com.kingdom.model.computer.ComputerPlayer
import com.kingdom.util.cardaction.CardActionHandler

import java.util.ArrayList

object KingdomComputerCardActionHandler {
    fun handleCardAction(oldCardAction: OldCardAction, computer: ComputerPlayer) {

        val game = computer.game
        val player = computer.player

        val cardName = oldCardAction.cardName
        val type = oldCardAction.type

        when (cardName) {
            "Artisan" -> when {
                oldCardAction.type == OldCardAction.TYPE_GAIN_CARDS_FROM_SUPPLY -> chooseHighestCostCard(oldCardAction, computer, game, player)
                else -> when {
                    player.actions == 0 && !player.actionCards.isEmpty() -> {
                        val card = computer.getHighestCostCard(player.actionCards)
                        chooseCard(oldCardAction, game, player, card)
                    }
                    else -> chooseLowestCostCard(oldCardAction, computer, game, player)
                }
            }
            "Bureaucrat" -> {
                val cardNames = computer.getCardsNotNeeded(oldCardAction.cards, 1)
                CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
            }
            "Cellar" -> {
                val cardsToDiscard = oldCardAction.cards
                        .filter { computer.isCardToDiscard(it) }
                        .map { it.name }
                CardActionHandler.handleSubmittedCardAction(game, player, cardsToDiscard, null, null, -1)
            }
            "Chancellor" -> CardActionHandler.handleSubmittedCardAction(game, player, null!!, "yes", null, -1)
            "Chapel" -> {
                val cardsToTrash = ArrayList<String>()
                for (card in oldCardAction.cards) {
                    if (computer.isCardToTrash(card)) {
                        cardsToTrash.add(card.name)
                    }
                    if (cardsToTrash.size == 4) {
                        break
                    }
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardsToTrash, null, null, -1)
            }
            "Feast" -> chooseHighestCostCard(oldCardAction, computer, game, player)
            "Library" -> CardActionHandler.handleSubmittedCardAction(game, player, null!!, "yes", null, -1)
            "Militia" -> {
                val numCardsToDiscard = player.hand.size - 3
                CardActionHandler.handleSubmittedCardAction(game, player, computer.getCardsToDiscard(oldCardAction.cards, numCardsToDiscard), null, null, -1)
            }
            "Mine" -> when (type) {
                OldCardAction.TYPE_TRASH_CARDS_FROM_HAND -> chooseLowestCostCard(oldCardAction, computer, game, player)
                else -> chooseHighestCostCard(oldCardAction, computer, game, player)
            }
            "Remodel" -> when (type) {
                OldCardAction.TYPE_TRASH_CARDS_FROM_HAND -> chooseLowestCostCard(oldCardAction, computer, game, player)
                else -> chooseHighestCostCard(oldCardAction, computer, game, player)
            }
            "Spy" -> {
                var yesNoAnswer = "yes"
                val topCard = oldCardAction.cards[0]
                if (topCard.name == Curse.NAME || topCard.isCopper) {
                    yesNoAnswer = "no"
                }
                player.getVictoryCards()
                        .filter { !it.isTreasure && !it.isAction }
                        .forEach { yesNoAnswer = "no" }
                if (oldCardAction.playerId == player.userId) {
                    if (yesNoAnswer == "yes") {
                        yesNoAnswer = "no"
                    } else {
                        yesNoAnswer = "yes"
                    }
                }
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, yesNoAnswer, null, -1)
            }
            "Thief" -> {
                val cardNames = ArrayList<String>()
                when (type) {
                    OldCardAction.TYPE_CHOOSE_CARDS -> when {
                        oldCardAction.numCards >= 1 -> {
                            val card1 = oldCardAction.cards[0]
                            when {
                                oldCardAction.cards.size == 2 -> {
                                    val card2 = oldCardAction.cards[1]
                                    when {
                                        card1.isTreasure && card2.isTreasure -> when {
                                            card1.cost > card2.cost -> cardNames.add(card1.name)
                                            else -> cardNames.add(card2.name)
                                        }
                                        card1.isTreasure -> cardNames.add(card1.name)
                                        else -> cardNames.add(card2.name)
                                    }
                                }
                                else -> cardNames.add(card1.name)
                            }
                        }
                    }
                    else -> when {
                        oldCardAction.numCards > 0 -> oldCardAction.cards
                                .filter { it.cost > 0 }
                                .mapTo(cardNames) { it.name }
                    }
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
            }
            "Throne Room" -> {
                val cardToPlay = computer.getActionToDuplicate(oldCardAction.cards, 2)
                chooseCard(oldCardAction, game, player, cardToPlay)
            }
            "Workshop" -> {
                val cardToGain = computer.getHighestCostCard(oldCardAction.cards)
                val cardNames = ArrayList<String>()
                cardNames.add(cardToGain!!.name)
                CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
            }
            else -> throw RuntimeException("Kingdom Card Action not handled for card: " + oldCardAction.cardName + " and type: " + oldCardAction.type)
        }
    }

    private fun chooseLowestCostCard(oldCardAction: OldCardAction, computer: ComputerPlayer, game: Game, player: Player) {
        val card = computer.getLowestCostCard(oldCardAction.cards)
        chooseCard(oldCardAction, game, player, card)
    }

    private fun chooseHighestCostCard(oldCardAction: OldCardAction, computer: ComputerPlayer, game: Game, player: Player) {
        val card = computer.getHighestCostCard(oldCardAction.cards)
        chooseCard(oldCardAction, game, player, card)
    }

    private fun chooseCard(oldCardAction: OldCardAction, game: Game, player: Player, card: Card?) {
        var cardToChoose = card
        if (cardToChoose == null) {
            cardToChoose = oldCardAction.cards[0]
        }
        val cardNames = ArrayList<String>()
        cardNames.add(cardToChoose.name)
        CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
    }
}
